### PR TITLE
add custom_ort_backend

### PR DESCRIPTION
**Pull request by tangyanf**
1. add onnxruntime custom op lib
2. add custom input data, because in some detection model such as: retinanet/faster-rcnn ..., if we use random input there will no anchor output, then we will encounter this error:
    2021-01-26 14:55:45.747470604 [E:onnxruntime:, sequential_executor.cc:333 Execute] Non-zero status code returned while running ReduceMax node. Name:'ReduceMax_1004' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/reduction/reduction_ops.cc:487 void onnxruntime::CommonReduce(onnxruntime::OpKernelContext*, std::vector<long int>, int64_t, onnxruntime::ResultsNoTransposePrepareForReduce&, bool) [with T = float; AGG = onnxruntime::ReduceAggregatorMax<float, float>; int64_t = long int] keepdims_ was false. Can't reduce on dim with value of 0 if 'keepdims' is false. Invalid output shape would be produced. input_shape:{0,4}

### CODE REVIEW

**Comment by daquexian** [comment]
Thanks for your great contribution! I'll check it these days

File: onnxsim/onnx_simplifier.py [hunk 1]
```diff
@@ -130,10 +131,13 @@ def is_dynamic(node):
...
 
 
-def forward(model, inputs=None, input_shapes: Optional[TensorShapes] = None) -> Dict[str, np.ndarray]:
+def forward(model, inputs: Optional[dict] = None, input_shapes: Optional[TensorShapes] = None,
+            custom_lib: Optional[str] = None) -> Dict[str, np.ndarray]:
     if input_shapes is None:
         input_shapes = {}
     sess_options = rt.SessionOptions()
+    if os.path.exists(custom_lib):
+        sess_options.register_custom_ops_library(custom_lib)
```
Comment (daquexian) [inline onnxsim/onnx_simplifier.py:140]: It will be better to raise an exception when the file doesn't exist

File: onnxsim/onnx_simplifier.py [hunk 2]
```diff
@@ -130,10 +131,13 @@ def is_dynamic(node):
     return copy.deepcopy(const_nodes)
 
 
-def forward(model, inputs=None, input_shapes: Optional[TensorShapes] = None) -> Dict[str, np.ndarray]:
+def forward(model, inputs: Optional[dict] = None, input_shapes: Optional[TensorShapes] = None,
```
Comment (daquexian) [inline onnxsim/onnx_simplifier.py:134]: How about creating a `Tensors = Dict[Optional[str], np.array]` in line 16 and replace the  `Optional[dict]` in this line and line 156 with `Optional[Tensors]`?

File: onnxsim/__main__.py [hunk 3]
```diff
@@ -18,26 +19,49 @@ def main():
     parser.add_argument('--skip-optimization', help='Skip optimization of ONNX optimizers.',
                         action='store_true')
     parser.add_argument(
-        '--input-shape', help='The manually-set static input shape, useful when the input shape is dynamic. The value should be "input_name:dim0,dim1,...,dimN" or simply "dim0,dim1,...,dimN" when there is only one input, for example, "data:1,3,224,224" or "1,3,224,224". Note: you might want to use some visualization tools like netron to make sure what the input name and dimension ordering (NCHW or NHWC) is.', type=str, nargs='+')
+        '--input-shape', help='The manually-set static input shape, useful when the input shape is dynamic. The value should be "input_name1:dim0,dim1,...,dimN input_name2:dim0,dim1...dimN", for example, "data1:1,3,224,224 data2:1,3,224,224". Note: you might want to use some visualization tools like netron to make sure what the input name and dimension ordering (NCHW or NHWC) is.', type=str, nargs='+')
```
Comment (daquexian) [inline onnxsim/__main__.py:22]: 'or simply "dim0,dim1,...,dimN"' is discarded. Is it intended?

File: onnxsim/__main__.py [hunk 4]
```diff
@@ -18,26 +19,49 @@ def main():
...
     parser.add_argument('--skip-shape-inference',
                         help='Skip shape inference. Shape inference causes segfault on some large models', action='store_true')
+    parser.add_argument(
+        '--input_data', help='input data, The value should be "input_name1:xxx1.bin  input_name2:xxx2.bin ...", input data should be a binary data file.', type=str, nargs='+')
+    parser.add_argument(
+        '--custom_lib', help="custom lib, if you have custom onnxruntime backend you should use this to register you custom op", type=str)
     args = parser.parse_args()
     print("Simplifying...")
-    input_shapes = {}
+    input_shapes = dict()
```
Comment (daquexian) [inline onnxsim/__main__.py:33]: Why use `dict()` instead of `{}`? :)

**Comment by daquexian** [comment]
Thanks for your great contribution! Overall it looks good to me. A problem is there is another pr about custom input data https://github.com/daquexian/onnx-simplifier/pull/61 which I will merge soon. Could you please update your pr (so that it only contains modifications about custom ort lib)?

**Comment by tangyanf** [comment]
> Thanks for your great contribution! Overall it looks good to me. A problem is there is another pr about custom input data #61 which I will merge soon. Could you please update your pr (so that it only contains modifications about custom ort lib)?
no problem，i will update it after you merge ：）

File: onnxsim/__main__.py (hunk 4 above)
Comment (tangyanf) [reply onnxsim/__main__.py:33]: {} is confused, if it is a set or a dict, so i used dict()

File: onnxsim/__main__.py (hunk 3 above)
Comment (tangyanf) [reply onnxsim/__main__.py:22]: sorry, it's my mistake, i delete it just want to match input_shape and input_data with the same input name, i forgot this option is optional

File: onnxsim/onnx_simplifier.py (hunk 1 above)
Comment (tangyanf) [reply onnxsim/onnx_simplifier.py:140]: great, i'll add an exception in there

File: onnxsim/onnx_simplifier.py (hunk 2 above)
Comment (tangyanf) [reply onnxsim/onnx_simplifier.py:134]: sure, that's great

**Comment by daquexian** [comment]
Hi @tangyanf , I have merged #61 about custom input data. Could you please update your PR so that only custom lib part is included? Thanks!

**Comment by tangyanf** [comment]
> Hi @tangyanf , I have merged #61 about custom input data. Could you please update your PR so that only custom lib part is included? Thanks!

hi@daquexian , i have migrated this pr to #121

**Pull request closed without merging**